### PR TITLE
Telemetry events for SKU Recommendation

### PR DESCRIPTION
**Pull request by neetusingh-ms**
Changes include -
1) Modified assessment telemetry to remove properties like subscription/resource group. This is needed because we moved azure login page just before migration and this information is not avaialble during assessment/sku recommendation.
2) Added following events -
    a. StartDataCollection - (when data collection is started)
            SessionId, TimeDataCollectionStarted.
image

    b. StopDataCollection - (when data collection is stopped)
            SessionId, TimeDataCollectionStopped.
image

    c. GetMISKURecommendation - (when MI SKU recommendation is generated).
            SessionId, targetMISKU.
image

    d. GetVMSKURecommendation -
            SessionId, targetVMSKU.
image

This is how recommendedSKU looks like -
image

    e.  GetInstanceRequirements -
image

This is how database level requirement looks like -
image
database level/ file level requirement contains database name/file name which has been hashed. No other PII information.

SessionId is to connect sku recommendation telemetry with assessment/ migration telemetry.

### CODE REVIEW

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 1]
```diff
@@ -387,12 +387,95 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+								};
+							})
+						};
+					}))
+				},
+				{}
+			);
+
+		} catch (e) {
+			console.log('error during get sku recommendation telemetry:');
```
Comment (rchlkm) [inline extensions/sql-migration/src/models/stateMachine.ts:474]: can we remove this log? we want to limit the number of prints

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 2]
```diff
@@ -387,12 +387,95 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+					},
+					{}
+				);
+			});
+
+			// Send Instance requirements used for calculating recommendations
+			sendSqlMigrationActionEvent(
+				TelemetryViews.SkuRecommendationWizard,
+				TelemetryAction.GetInstanceRequirements,
+				{
```
Comment (rchlkm) [inline extensions/sql-migration/src/models/stateMachine.ts:429]: can we add the performance data source to track if user was getting the recommendation using "collect data" or "open existing"?
`performanceDataSource = this._skuRecommendationPerformanceDataSource`

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 3]
```diff
@@ -531,9 +655,6 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 				TelemetryAction.ServerAssessment,
 				{
 					'sessionId': this._sessionId,
-					'tenantId': this._azureAccount.properties.tenants[0].id,
```
Comment (rchlkm) [inline extensions/sql-migration/src/models/stateMachine.ts:534]: Even though we're removing the tenantId, subscriptionId, resourceGroup during generateAssessment(), we still need this information. Can we add these data points on the targetSelectionPage?

We can send the telemetry onPageLeave once the user selects these values.
https://github.com/microsoft/azuredatastudio/blob/854e90b772537abc677e7e54d01124c8e4d18c2d/extensions/sql-migration/src/wizard/targetSelectionPage.ts#L131

**Review by rchlkm — CHANGES_REQUESTED** [review]

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 4]
```diff
@@ -468,9 +572,29 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
 	}
 
+	private async generateStopDataCollectionTelemetry(): Promise<void> {
+		try {
+			sendSqlMigrationActionEvent(
+				TelemetryViews.DataCollectionWizard,
+				TelemetryAction.StopDataCollection,
+				{
+					'sessionId': this._sessionId,
+					'TimeDataCollectionStopped': this._perfDataCollectionStopDate?.toString()
```
Comment (rchlkm) [inline extensions/sql-migration/src/models/stateMachine.ts:587]: nit: can we lowercase to `timeDataCollectionStopped` to keep the formatting consistent

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 5]
```diff
@@ -387,12 +387,95 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+		try {
+
+			this._skuRecommendationResults?.recommendations?.sqlMiRecommendationResults?.forEach(resultItem => {
+				// Send telemetry for recommended MI SKU
+				sendSqlMigrationActionEvent(
+					TelemetryViews.SkuRecommendationWizard,
+					TelemetryAction.GetMISkuRecommendation,
+					{
+						'sessionId': this._sessionId,
+						'RecommendedSku': JSON.stringify(resultItem?.targetSku)
```
Comment (rchlkm) [inline extensions/sql-migration/src/models/stateMachine.ts:406]: nit: please camel case to `recommendedSku`

File: extensions/sql-migration/src/models/stateMachine.ts (hunk 3 above)
Comment (neetusingh-ms) [reply extensions/sql-migration/src/models/stateMachine.ts:534]: You mean just the tenantId,subid, rg? or entire assessment telemetry.

Comment (neetusingh-ms) [reply extensions/sql-migration/src/models/stateMachine.ts:534]: Done

**Review by rchlkm — APPROVED** [review]

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 6]
```diff
@@ -394,12 +394,95 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+			sendSqlMigrationActionEvent(
+				TelemetryViews.SkuRecommendationWizard,
+				TelemetryAction.GetInstanceRequirements,
+				{
+					'sessionId': this._sessionId,
+					'performanceDataSource': this._skuRecommendationPerformanceDataSource,
+					'cpuRequirementInCores': this._skuRecommendationResults?.recommendations?.instanceRequirements?.cpuRequirementInCores?.toString(),
+					'dataStorageRequirementInMB': this._skuRecommendationResults?.recommendations?.instanceRequirements?.dataStorageRequirementInMB?.toString(),
+					'logStorageRequirementInMB': this._skuRecommendationResults?.recommendations?.instanceRequirements?.logStorageRequirementInMB?.toString(),
+					'memoryRequirementInMB': this._skuRecommendationResults?.recommendations?.instanceRequirements?.memoryRequirementInMB.toString(),
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:442]: Generally number values like this should go in the measurements fields

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 7]
```diff
@@ -394,12 +394,95 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+							ioThroughputRequirementInMBps: i.ioThroughputRequirementInMBps,
+							dataStorageRequirementInMB: i.dataStorageRequirementInMB,
+							logStorageRequirementInMB: i.logStorageRequirementInMB,
+							databaseName: hashString(i.databaseName),
+							memoryRequirementInMB: i.memoryRequirementInMB,
+							cpuRequirementInPercentageOfTotalInstance: i.cpuRequirementInPercentageOfTotalInstance,
+							numberOfDataPointsAnalyzed: i.numberOfDataPointsAnalyzed,
+							fileLevelRequirements: i.fileLevelRequirements?.map(file => {
+								return {
+									fileName: hashString(file.fileName),
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:465]: You're hashing this, which is good. But do you really need this? Generally we prefer that unless you actually have a use case for a data point that you just leave it out.

File: extensions/sql-migration/src/wizard/targetSelectionPage.ts [hunk 8]
```diff
@@ -130,6 +131,17 @@ export class TargetSelectionPage extends MigrationWizardPage {
...
 		this.wizard.registerNavigationValidator((e) => {
+
+			sendSqlMigrationActionEvent(
+				TelemetryViews.MigrationWizardTargetSelectionPage,
+				TelemetryAction.OnPageLeave,
+				{
+					'sessionId': this.migrationStateModel?._sessionId,
+					'subscriptionId': this.migrationStateModel?._targetSubscription?.id,
+					'resourceGroup': this.migrationStateModel?._resourceGroup?.name,
+					'tenantId': this.migrationStateModel?._accountTenants[0]?.id
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/wizard/targetSelectionPage.ts:142]: Is grabbing the first one guaranteed to be correct?

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/sql-migration/src/wizard/targetSelectionPage.ts (hunk 8 above)
Comment (neetusingh-ms) [reply extensions/sql-migration/src/wizard/targetSelectionPage.ts:142]: @rachel-msft  can you pls confirm? this is how we were doing before.

Comment (raymondtruong) [reply extensions/sql-migration/src/wizard/targetSelectionPage.ts:142]: @rchlkm

File: extensions/sql-migration/src/models/stateMachine.ts (hunk 7 above)
Comment (neetusingh-ms) [reply extensions/sql-migration/src/models/stateMachine.ts:465]: Thanks Charles for pointing this.. we don't actually need file names, filetype is important. Let me remove in next commit.

**Review by rchlkm — APPROVED** [review]

**Pull request merged**